### PR TITLE
[CDAP-12109] Ability to add arbitrary branching of the Workflow fork and conditions in the pipeline.

### PR DESCRIPTION
**Pull request by sagarkapare**
This PR depends on #9363 only for supporting Condition api in the Workflow, which is used in pipeline. This PR contains support for dynamically configuring the multi level conditions and forks in the pipeline.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToConditionAdder.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToConditionAdder.java]: I think it is a little cleaner to do:

```
configurer.otherwise();
return this;
```

This would also make it clearer to people looking at this in the future that they should not treat this object and the object returned by this method any differently.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToForkAdder.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToForkAdder.java]: similar comment about doing this as:

```
configurer.otherwise();
return this;
```

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToTrunkAdder.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToTrunkAdder.java]: similar comment here

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ForkToConditionAdder.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ForkToConditionAdder.java]: similar comment here

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ForkToForkAdder.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ForkToForkAdder.java]: similar comment

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ForkToTrunkAdder.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ForkToTrunkAdder.java]: similar comment

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToConditionAdder.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToConditionAdder.java]: Is there a reason we need T instead of just always using WorkflowProgramAdder as the parent type?

Comment (sagarkapare) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToConditionAdder.java]: No need for this to be typed parameter. I started with the configurer as a parent but later realize that we need to send the entire object so changed it back to WorkflowProgramAdder. fixed this now.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToConditionAdder.java
Comment (sagarkapare) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/ConditionToConditionAdder.java]: makes sense. updated in all places.

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by sagarkapare** [comment]
Release build is running here - https://builds.cask.co/browse/CDAP-RUT1220-1

**Comment by sagarkapare** [comment]
Build is green here - https://builds.cask.co/browse/CDAP-RUT1220-1. Will merge once the #9363 gets merged.

**Comment by sagarkapare** [comment]
Build is already green. Squashing and merging.

**Pull request merged**
